sycl/source/virtual_mem.cpp

```diff
@@ -0,0 +1,156 @@
+
+#include <detail/context_impl.hpp>
+#include <detail/device_impl.hpp>
+#include <detail/physical_mem_impl.hpp>
+#include <sycl/ext/oneapi/virtual_mem/virtual_mem.hpp>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext::oneapi::experimental {
+
+__SYCL_EXPORT size_t get_minimum_mem_granularity(const device &SyclDevice,
+                                                 const context &SyclContext) {
+  if (!SyclDevice.has(aspect::ext_oneapi_virtual_mem))
+    throw sycl::exception(
+        sycl::make_error_code(sycl::errc::feature_not_supported),
+        "Device does not support aspect::ext_oneapi_virtual_mem.");
+
+  std::shared_ptr<sycl::detail::device_impl> DeviceImpl =
+      sycl::detail::getSyclObjImpl(SyclDevice);
+  std::shared_ptr<sycl::detail::context_impl> ContextImpl =
+      sycl::detail::getSyclObjImpl(SyclContext);
+  const sycl::detail::plugin &Plugin = ContextImpl->getPlugin();
+#ifndef NDEBUG
+  size_t InfoOutputSize;
+  Plugin.call<sycl::detail::PiApiKind::piextVirtualMemGranularityGetInfo>(
+      ContextImpl->getHandleRef(), DeviceImpl->getHandleRef(),
+      PI_EXT_ONEAPI_VIRTUAL_MEM_GRANULARITY_INFO_MINIMUM, 0, nullptr,
+      &InfoOutputSize);
+  assert(InfoOutputSize == sizeof(size_t) &&
+         "Unexpected output size of granularity info query.");
+#endif // NDEBUG
+  size_t Granularity;
+  Plugin.call<sycl::detail::PiApiKind::piextVirtualMemGranularityGetInfo>(
+      ContextImpl->getHandleRef(), DeviceImpl->getHandleRef(),
+      PI_EXT_ONEAPI_VIRTUAL_MEM_GRANULARITY_INFO_MINIMUM, sizeof(size_t),
+      &Granularity, nullptr);
+  return Granularity;
+}
+
+__SYCL_EXPORT size_t get_recommended_mem_granularity(
+    const device &SyclDevice, const context &SyclContext) {
+  if (!SyclDevice.has(aspect::ext_oneapi_virtual_mem))
+    throw sycl::exception(
+        sycl::make_error_code(sycl::errc::feature_not_supported),
+        "Device does not support aspect::ext_oneapi_virtual_mem.");
+
+  std::shared_ptr<sycl::detail::device_impl> DeviceImpl =
+      sycl::detail::getSyclObjImpl(SyclDevice);
+  std::shared_ptr<sycl::detail::context_impl> ContextImpl =
+      sycl::detail::getSyclObjImpl(SyclContext);
+  const sycl::detail::plugin &Plugin = ContextImpl->getPlugin();
+#ifndef NDEBUG
+  size_t InfoOutputSize;
+  Plugin.call<sycl::detail::PiApiKind::piextVirtualMemGranularityGetInfo>(
+      ContextImpl->getHandleRef(), DeviceImpl->getHandleRef(),
+      PI_EXT_ONEAPI_VIRTUAL_MEM_GRANULARITY_INFO_RECOMMENDED, 0, nullptr,
+      &InfoOutputSize);
+  assert(InfoOutputSize == sizeof(size_t) &&
+         "Unexpected output size of granularity info query.");
+#endif // NDEBUG
+  size_t Granularity;
+  Plugin.call<sycl::detail::PiApiKind::piextVirtualMemGranularityGetInfo>(
+      ContextImpl->getHandleRef(), DeviceImpl->getHandleRef(),
+      PI_EXT_ONEAPI_VIRTUAL_MEM_GRANULARITY_INFO_RECOMMENDED, sizeof(size_t),
+      &Granularity, nullptr);
+  return Granularity;
```

Comment:
I see that the only difference in minimum and recommended granularity query implementation is info id. Is it feasible to extract this common logic to a function accepting PI_EXT_ONEAPI_VIRTUAL_MEM_GRANULARITY_INFO_RECOMMENDED/MINIMUM as well?

Reply:
As a user-side interface? So maybe something like:

```
enum class granularity_mode { minimum, recommended };

size_t get_mem_granularity(const device &SyclDevice,
                                                 const context &SyclContext,
                                                 granularity_mode Mode = granularity_mode::minimum);
```

Whether or not there should be a default for `Mode` (and whether it should be `minimum`) I'm not sure.

Reply:
If we will go this way, I think we should choose recommended mode as default.

Reply:
Yes, I like it!  I assume the default should be "recommended", not "minimum", right?

Reply:
Yes, it should probably default to recommended. I have made the corresponding changes.